### PR TITLE
[SPARK-34306][SQL][PYTHON][R] Use Snake naming rule across the function APIs

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR completes snake_case rule at functions APIs across the languages, see also SPARK-10621.

In more details, this PR:
- Adds `count_distinct` in Scala Python, and R, and document that `count_distinct` is encouraged. This was not deprecated because `countDistinct` is pretty commonly used. We could deprecate in the future releases.
- (Scala-specific) adds `typedlit` but doesn't deprecate `typedLit` which is arguably commonly used. Likewise, we could deprecate in the future releases.
- Deprecates and renames:
  - `sumDistinct` -> `sum_distinct`
  - `bitwiseNOT` -> `bitwise_not`
  - `shiftLeft` -> `shiftleft` (matched with SQL name in `FunctionRegistry`)
  - `shiftRight` -> `shiftright` (matched with SQL name in `FunctionRegistry`)
  - `shiftRightUnsigned` -> `shiftrightunsigned` (matched with SQL name in `FunctionRegistry`)
  - (Scala-specific) `callUDF` -> `call_udf`

### Why are the changes needed?

To keep the consistent naming in APIs.

### Does this PR introduce _any_ user-facing change?

Yes, it deprecates some APIs and add new renamed APIs as described above.

### How was this patch tested?

Unittests were added.

### CODE REVIEW

File: python/docs/source/reference/pyspark.sql.rst [hunk 1]
```diff
@@ -482,9 +484,9 @@ Functions
     sequence
     sha1
     sha2
-    shiftLeft
-    shiftRight
-    shiftRightUnsigned
+    shiftleft
+    shiftright
+    shiftrightunsigned
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql.rst:489]: I should rename it here - the case sensitivity is not taken into account `sphinx.ext.autosummary` due to the file name conflicts at RST stubs.

**Comment by HyukjinKwon** [comment]
cc @rxin, @srowen, @zero323, and @viirya FYI. Could you take a look please when you guys find some time.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 2]
```diff
@@ -123,9 +133,9 @@ object functions {
    * can handle parameterized scala types e.g.: List, Seq and Map.
    *
    * @group normal_funcs
-   * @since 2.2.0
+   * @since 3.2.0
    */
-  def typedLit[T : TypeTag](literal: T): Column = literal match {
+  def typedlit[T : TypeTag](literal: T): Column = literal match {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:138]: not `typed_lit` but `typedlit`?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 3]
```diff
@@ -113,6 +113,16 @@ object functions {
    */
   def lit(literal: Any): Column = typedLit(literal)
 
+  /**
+   * Creates a [[Column]] of literal value.
+   *
+   * An alias of `typedlit`, and it is encouraged to use `typedlit` directly.
+   *
+   * @group normal_funcs
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:121]: Could we use `@deprecated` for these aliases?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:138]: For some reasons I like `typedlit` (like `datediff`, `shiftleft, etc.). But I don't mind changing it back if other people prefer `typed_lit` more. It's not even in the function registry in any event so I think we can pick one.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 3 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:121]: oh I didn't intend to deprecate this (and `count_dictinct`). I will fix PR description.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 2 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:138]: Ah, I see. I just wanted to know the reason. The current one looks fine.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 3 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:121]: Oh, I see. Yea, I misunderstood it from the word "deprecate" there.

**Review by viirya — COMMENTED** [review]
Are these the all functions we need to change? Overall I think it is good to keep consistent naming across languages. For the Scala-specific functions, is it necessary to change too?

**Comment by HyukjinKwon** [comment]
Yeah, these are everything if I didn't miss. I could take the Scala specific ones out but just thought it's better to keep it consistent within `functions.scala` too, and they are only two (`callUDF` and `typedLit`). I don't mind changing these two back if any of you feel it shouldn't be changed.

**Review by srowen — COMMENTED** [review]
Seems OK for better consistency, as long as nothing is removed now.

File: R/pkg/R/functions.R [hunk 4]
```diff
@@ -636,20 +636,33 @@ setMethod("bin",
...
+#'
+#' @rdname column_nonaggregate_functions
+#' @aliases bitwiseNOT bitwiseNOT,Column-method
 #' @note bitwiseNOT since 1.5.0
 setMethod("bitwiseNOT",
           signature(x = "Column"),
           function(x) {
-            jc <- callJStatic("org.apache.spark.sql.functions", "bitwiseNOT", x@jc)
-            column(jc)
+            .Deprecated("bitwise_not")
```
Comment (dongjoon-hyun) [inline R/pkg/R/functions.R:664]: Shall we not deprecate this because Apache Spark cannot delete this in the future?

**Review by dongjoon-hyun — COMMENTED** [review]
+1 for the idea with a question about about the necessity of deprecation.

File: R/pkg/R/functions.R (hunk 4 above)
Comment (srowen) [reply R/pkg/R/functions.R:664]: That's a reasonable position too. You _could_ in 4.0, but, dont' imagine anyone would consider it worth breaking.

**Comment by rxin** [comment]
This makes sense, assuming we don't break backward compatibility.

File: R/pkg/R/functions.R (hunk 4 above)
Comment (HyukjinKwon) [reply R/pkg/R/functions.R:664]: I think deprecation is fine because it guides users to use the encouraged way.

**Comment by HyukjinKwon** [comment]
Thanks all. Merged to master.

**Pull request closed without merging**